certificate-rotation.md

```diff
@@ -0,0 +1,113 @@
+There are two parts to certificate rotation: generating the new certificates, and reloading them.
+
+Both have been improved, with the main difference being that we can now reload certificates without restarting a node.
+
+## Order of certificate regeneration
+
+Most of the complication in rotating certificates is proper handling of new CA certificates.
+It's important for all parts of the system (nodes and clients) to have the same CA certificates, or they may not be able to verify each other's client/node certificates.
+
+It is possible to have multiple CA certificates active at the same time, but not node or client certificates. Therefore, CA certificates should be generated and pushed first, followed by new node/client certificates using the new CA.
+
+Let's say we want to regenerate all certificates (CA, client, node), we should do the following:
+1. generate new CA certificate and combine it with the existing one
+1. push combined CA certificate to all nodes and reload certificates
+1. make combined CA available to all clients. We may need to wait some time (days/weeks) for all clients to make use of the new certificate
+1. generate and push new node/client certificates using the new CA certificate
+
+It is recommended to change CA certificates on a completely different schedule. eg: months before.
+**WARNING**: node certificates generated with the new CA will not be accepted by clients that do not yet have the new CA. Before all clients have the new CA cert, node certificates should be signed with the old CA (this is still a [TODO](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/cert.go#L91-L93))
+
+## Generating new CA certificates
+
+CA rotation includes generating a new CA certificate and key.
+
+If the original certificate and key are generated using:
+```
+$ cockroach cert create-ca --certs-dir=certs --ca-key=ca.key --logtostderr=true
+I170425 14:20:29.394473 1 security/certs.go:120  Generated CA key ca.key
+I170425 14:20:29.398081 1 security/certs.go:173  Wrote 1 certificates to certs/ca.crt
+```
+
+We can generate a new CA certificate and key with the following:
+```
+$ cockroach cert create-ca --certs-dir=certs --ca-key=ca.new.key --overwrite --logtostderr=true
+I170425 14:21:10.490341 1 security/certs.go:120  Generated CA key ca.new.key
+I170425 14:21:10.494007 1 security/certs.go:160  Found 1 certificates in certs/ca.crt
+I170425 14:21:10.494341 1 security/certs.go:173  Wrote 2 certificates to certs/ca.crt
+```
+
+Flags:
+* `--ca-key=ca.new.key`: path to the new key to generate. To re-use the existing key (not recommended), use `--ca-key=ca.key` and `--allow-key-reuse`.
```

Comment:
@mberhault In which scenario would we need to reuse an existing key?

Reply:
Reusing the old key is possible, just not recommended. For security reasons, it is considered good practice to rotate keys on a regular basis (that's the sensitive part). Someone could ignore the recommendation entirely for any reason they like (one less file to keep and distribute)